src/python/pants/backend/scala/compile/scalac.py

```diff
@@ -184,33 +191,43 @@ async def compile_scala_source(
                 scala_artifacts.compiler_main,
                 "-bootclasspath",
                 ":".join(tool_classpath.classpath_entries(toolcp_relpath)),
-                *local_plugins.args(local_scalac_plugins_relpath),
+                *scalac_plugins.args(scalac_plugins_relpath),
                 *(("-classpath", classpath_arg) if classpath_arg else ()),
                 *scalac.args,
-                # NB: We set a non-existent main-class so that using `-d` produces a `jar` manifest
-                # with stable content.
-                "-Xmain-class",
-                NO_MAIN_CLASS,
                 "-d",
-                output_file,
+                output_dir,
                 *sorted(
                     chain.from_iterable(
                         sources.snapshot.files
                         for _, sources in component_members_and_scala_source_files
                     )
                 ),
             ],
-            input_digest=sources_digest,
+            input_digest=merged_digest,
             extra_immutable_input_digests=extra_immutable_input_digests,
             extra_nailgun_keys=extra_nailgun_keys,
-            output_files=(output_file,),
+            output_directories=(output_dir,),
             description=f"Compile {request.component} with scalac",
             level=LogLevel.DEBUG,
         ),
     )
+
     output: ClasspathEntry | None = None
     if process_result.exit_code == 0:
-        output_digest = process_result.output_digest
+        compilation_results = await Get(
+            Snapshot, RemovePrefix(process_result.output_digest, output_dir)
+        )
+        logging.debug(f"Scalac compilation results: {compilation_results.files}")
+        output_digest = await Get(
+            Digest,
+            JarToolRequest(
+                jar_name=output_file,
+                digest=compilation_results.digest,
+                main_class=NO_MAIN_CLASS,
+                files=compilation_results.files,
+            ),
```

Comment:
Because the output of `scalac` is no longer a `jar` file, we build it ourselves now.

Reply:
Maybe move this scalac compilation change to its own PR?